### PR TITLE
[DRAFT] Re-add #9202, but with less crashes

### LINKED ISSUE #10034: Terminal crashes when running git log
### Windows Terminal version (or Windows build number)

Windows Build: 10.0.19042.964, Windows Terminal commit c3ca94cecaef5a9b22e94a7c34983b19cf77e252 built using Visual Studio 2019 16.9.3

### Other Software

PowerShell 7.1.3
oh-my-posh 3.133.1
posh-git 1.0.0
git 2.31.1.windows.1

### Steps to reproduce

- Start any shell with access to git (powershell, git bash, cmd)
- git clone the windows terminal repo
- git checkout master
- git log
- press space enough times to see this screen, one more space will kill it depending on the number of lines in the terminal

### Expected Behavior

git log shouldn't cause wtd to crash.

### Actual Behavior

image

Press space after this screen and it wtd will crash leaving the following windows event log entry.

Fault bucket 1724119567131471671, type 5
Event Name: MoAppCrash
Response: Not available
Cab Id: 0

Problem signature:
P1: WindowsTerminalDev_0.0.1.0_x64__8wekyb3d8bbwe
P2: praid:App
P3: 0.0.0.0
P4: 6091d600
P5: DWrite.dll
P6: 10.0.19041.804
P7: 2f618cb5
P8: c0000005
P9: 00000000000cf34e
P10: 

Attached files:
\\?\WERD4E7.tmp.mdmp
\\?\WERD630.tmp.WERInternalMetadata.xml
\\?\C:\ProgramData\Microsoft\Windows\WER\Temp\WERD641.tmp.xml
\\?\C:\ProgramData\Microsoft\Windows\WER\Temp\WERD64F.tmp.csv
\\?\C:\ProgramData\Microsoft\Windows\WER\Temp\WERD67E.tmp.txt

These files may be available here:
\\?\C:\ProgramData\Microsoft\Windows\WER\ReportArchive\AppCrash_WindowsTerminalD_14915a2eb4f1c128f427ee9d7511ba25d013a0ff_a3463a88_f44a4158-0e6d-4109-867f-3e07b7600941

Analysis symbol: 
Rechecking for solution: 0
Report Id: b377ecbe-9dab-4a05-beea-55614e8d7bb0
Report Status: 268435456
Hashed bucket: c432aee12ffbe1ec97ed4d9c8d2bef37
Cab Guid: 0
 
[CrashDetails.zip](https://github.com/microsoft/terminal/files/6424622/CrashDetails.zip)

[settings.zip](https://github.com/microsoft/terminal/files/6424649/settings.zip)

### PR DESCRIPTION
**Pull request by skyline75489**
## Summary of the Pull Request

Introduced by #9202, I mistakenly used the `glyphStart` instead of `textStart` to calculate the cluster map of simple runs, causing the application to crash when it hits N-M cluster maps. This PR corrects the mistake.

 
## PR Checklist
* [X] Closes #10034
* [ ] CLA signed. If not, go over [here](https://cla.opensource.microsoft.com/microsoft/Terminal) and sign the CLA
* [ ] Tests added/passed
* [ ] Documentation updated. If checked, please file a pull request on [our docs repo](https://github.com/MicrosoftDocs/terminal) and link it here: #xxx
* [ ] Schema updated.
* [ ] I've discussed this with core contributors already. If not checked, I'm ready to accept this work might be rejected in favor of a different grand plan. Issue number where discussion took place: #xxx

### CODE REVIEW

**Review by zadjii-msft — APPROVED** [review]

**Comment by skyline75489** [comment]
OK I found more defects in #9202 and I'm marking this as WIP. I'm OK with reverting #9202 because this PR is not pretty... 

CC and Requesting help from @miniksa

File: src/renderer/dx/CustomTextLayout.cpp
Comment (skyline75489) [reply src/renderer/dx/CustomTextLayout.cpp:194]: Even with the fix, the `_glyphIndices` for simple runs will have *wrong start* index, if there's N-M runs previous to them. For example a simple line like this:

```text
😀
```

There's two runs. First is complex, the second is simple. 

| Text | Length | Glyph Count |
| -- | -- | -- |
| 😀| 2 | 1 |
| (A lot of spaces | 118 | 118 |

The correct `_glyphIndices` is this:

| Index | Value |
| -- | -- |
| 0  | 11685 |
| 1  | 3 |
| 2  | 3 |
| 3 | 3 | 
| ... | ... |

But here we have:

| Index | Value |
| -- | -- |
| 0  | 11685 |
| 1  | **0** |
| 2  | 3 |
| 3 | 3 | 
| ... | ... |

Notice there's an incorrect 0 indice here, because when we do `GetTextComplexity` on the second run, we have no idea how many glyphs the previous run takes up, and assume the number is equal to the textLength (2) and fill up the `_glyphIndices` from there.

File: src/renderer/dx/CustomTextLayout.cpp
Comment (skyline75489) [inline src/renderer/dx/CustomTextLayout.cpp:454]: This is me trying to fix the issue mentioned above by left-shifting the entire array to avoid empty holes in `_glyphIndices`.

**Comment by miniksa** [comment]
If this is insufficient and doesn't fix the problems from #9202, then we need to revert it until we figure it out. I don't have the brain capacity to dive into this space this week. cc: @DHowett

**Comment by skyline75489** [comment]
This fixes the problems for #9202 (both crash & rendering issue) but I'm not confident that this is the proper way, or maybe this will leads to other unknown issues. So reverting #9202 might be a good option to unblock future releases.

**Comment by DHowett** [comment]
For now I'll revert 9502. Sorry @skyline75489  :smile:

**Comment by DHowett** [comment]
Sorry, 9202.

**Comment by zadjii-msft** [comment]
Should we repurpose this issue as "re-add #9202, but with less crashes", or should we open a new PR for that? I don't really care which ¯\\\_(ツ)_/¯

**Comment by DHowett** [comment]
> Should we repurpose this issue as "re-add #9202, but with less crashes", or should we open a new PR for that? I don't really care which ¯\_(ツ)_/¯

I'd like to repurpose it!

**Comment by skyline75489** [comment]
Hey guys I've repurposed this. I realized that in order to fully solve this, we might need to rewrite the underlying run structure. Is the DxRenderer included in the TextBuffer rewrite of 2021? I think this is somewhat blocked by that.

**Comment by DHowett** [comment]
It's not expected that the text buffer change will cause or require any changes in the renderer :) The interfaces through which the rest of the code reads the buffer are staying unchanged

**Comment by skyline75489** [comment]
Closed in favor of future work for #9156

**Pull request closed without merging**
